File: backend/app/models/documents.py
"""
Pydantic models for document upload and processing.
"""

from datetime import datetime
from typing import Optional, Literal

from pydantic import BaseModel


class DocumentChunk(BaseModel):
    """A single chunk of document content for indexing."""
    id: str
    source_type: Literal["pdf", "docx"]
    document_id: str
    title: str
    filename: str
    page_number: int
    section_heading: Optional[str] = None
    chunk_index: int
    content: str


class DocumentUploadResponse(BaseModel):
    """Response from document upload endpoint."""
    success: bool
    document_id: str
    filename: str
    chunks_created: int
    gcs_uri: str
    error: Optional[str] = None


class DocumentMetadata(BaseModel):
    """Document metadata stored in Firestore."""
    document_id: str
    filename: str
    title: str
    source_type: Literal["pdf", "docx"]
    page_count: int
    chunk_count: int
    upload_date: datetime
    gcs_original_uri: str
    gcs_jsonl_uri: str

File: frontend/lib/types/documents.ts
/**
 * Document upload and display types
 */

export interface DocumentUploadResponse {
  success: boolean
  document_id: string
  filename: string
  chunks_created: number
  gcs_uri: string
  error?: string
}

export interface DocumentMetadata {
  document_id: string
  filename: string
  title: string
  source_type: 'pdf' | 'docx'
  page_count: number
  chunk_count: number
  upload_date: string
  gcs_original_uri: string
  gcs_jsonl_uri: string
}

export interface DocumentSource {
  type: 'pdf' | 'docx'
  document_id: string
  title: string
  filename: string
  page_number: number
  section_heading?: string
  snippet: string
  deep_link?: string
}

export interface YouTubeSource {
  type: 'youtube'
  title: string
  url: string
  video_id: string
  timestamp: number
  snippet: string
}

// Union type for all source types
export type UnifiedSource = YouTubeSource | DocumentSource

// Type guard for YouTube sources
export function isYouTubeSource(source: UnifiedSource): source is YouTubeSource {
  return source.type === 'youtube'
}

// Type guard for document sources
export function isDocumentSource(source: UnifiedSource): source is DocumentSource {
  return source.type === 'pdf' || source.type === 'docx'
}
